src/Common/AssemblyLoadingUtils.cs

```diff
@@ -0,0 +1,249 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.ML.Runtime.Internal.Utilities;
+using System;
+using System.IO;
+using System.IO.Compression;
+using System.Reflection;
+
+namespace Microsoft.ML.Runtime
+{
+    internal static class AssemblyLoadingUtils
+    {
+        /// <summary>
+        /// Make sure the given assemblies are loaded and that their loadable classes have been catalogued.
+        /// </summary>
+        public static void LoadAndRegister(IHostEnvironment env, string[] assemblies)
+        {
+            Contracts.AssertValue(env);
+
+            if (Utils.Size(assemblies) > 0)
+            {
+                foreach (string path in assemblies)
+                {
+                    Exception ex = null;
+                    try
+                    {
+                        // REVIEW: Will LoadFrom ever return null?
+                        Contracts.CheckNonEmpty(path, nameof(path));
+                        var assem = LoadAssembly(env, path);
+                        if (assem != null)
+                            continue;
+                    }
+                    catch (Exception e)
+                    {
+                        ex = e;
+                    }
+
+                    // If it is a zip file, load it that way.
+                    ZipArchive zip;
+                    try
+                    {
+                        zip = ZipFile.OpenRead(path);
+                    }
+                    catch (Exception e)
+                    {
+                        // Couldn't load as an assembly and not a zip, so warn the user.
+                        ex = ex ?? e;
+                        Console.Error.WriteLine("Warning: Could not load '{0}': {1}", path, ex.Message);
+                        continue;
+                    }
+
+                    string dir;
+                    try
+                    {
+                        dir = CreateTempDirectory();
+                    }
+                    catch (Exception e)
+                    {
+                        throw Contracts.ExceptIO(e, "Creating temp directory for extra assembly zip extraction failed: '{0}'", path);
+                    }
+
+                    try
+                    {
+                        zip.ExtractToDirectory(dir);
+                    }
+                    catch (Exception e)
+                    {
+                        throw Contracts.ExceptIO(e, "Extracting extra assembly zip failed: '{0}'", path);
+                    }
+
+                    LoadAssembliesInDir(env, dir, false);
+                }
+            }
+        }
+
+        public static IDisposable CreateAssemblyRegistrar(IHostEnvironment env, string loadAssembliesPath = null)
```

Comment:
So the only thing that can be done with the registrar is to dispose it?

Reply:
Correct. This is an internal method that is only invoked on that command-line. It creates an object that listens for new assemblies to be loaded, and registers them automatically.